Add BlockDataManager and admin tools and complete frying pan and trash can functions

BrodBlox09 opened the pull request:
BlockDataManager controls the data of blocks and block data stored in items. Frying pan now accepts items and holds them for later use (use of items not yet implemented). Trash can now clears the items in a frying pan (does not take into account whether or not the items it clears are "burnt" or not, functionality may need to be updated when there is a way to tell if an item is burnt). Admin tools have been added to aid in the development process (e.g. checking to see if block data is being stored properly).

BrodBlox09 committed:
Add more methods to Vector helper class

BrodBlox09 committed:
Add BlockDataManager

BrodBlox09 committed:
Organize BlockDataManager

BrodBlox09 committed:
Add functionality to BlockDataManager

Improve type-safety and ease-of-use (getters and setters now use Block objects by default instead of Vector locations) and add the ability to store/retrieve block information to/from item stacks.

BrodBlox09 committed:
Add admin tools

To test features more easily, add admin tools. Currently, it can do the following:
"bcc.cook:getBlockData":
Retrieves the block data of the block the player is currently looking at.
"bcc.cook:getSelectedItemBlockData":
Retrieves the block data of the selected item.

BrodBlox09 committed:
Add frying pan block data interface

BrodBlox09 committed:
Add custom frying pan item

Instead of using Minecraft's automatically generated item for the frying pan, add a custom frying pan item that is not stackable to be able to use item dynamic properties.

BrodBlox09 committed:
Add 2 methods to PlayerInventory helper class

clearItem and setItem

BrodBlox09 committed:
Update frying pan functionality

Frying pan can now have fryable items deposited into it. When picked up, the information of these items stay tracked on the frying pan item. When implemented, the player will be able to see this information. For now, admin tools should be used.

BrodBlox09 committed:
Add custom frying pan functions to trash bin

Now, when the trash bin is interacted with the frying pan, information about what the frying pan is holding is cleared, effectively clearing the food on it. The Miro board specifies that the trash bin clears burnt food, but currently there is no way to tell if an item is considered "burnt" or not. Functionality may need to change in the future for this reason.

BrodBlox09 committed:
Add missing licensing headers to script files

BrodBlox09 committed:
Ensure block data is cleared when block is broken

BrodBlox09 committed:
Merge branch 'main' into fryingPan

BrodBlox09 commented:
Merge conflicts resolved

jeanmajid commented on behavior_pack/src/adminTools.ts:
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: MIT
```
Lets move admin tools under components/items/admin or something like that, theres no need to make a whole different system for just the admin tools. Also no need for any tag checking, if the player is able to get the admin tool in the first place, they already put themselves into creative mode anyways

BrodBlox09 replied:
Resolved in 46142c702bada73332c03b47d31c75c9b8a1b88f

jeanmajid commented on behavior_pack/src/PlayerInventory.ts:
```diff
@@ -23,7 +23,15 @@ export default class PlayerInventory {
 
-	public hasEmptySlot() {
+	public clearItem(slotIndex: number): void {
+		this.setItem(slotIndex);
```
might be a bit nitpicky, so its not a must, but specify undefined or null as the second argument, it makes it more readable

BrodBlox09 replied:
Resolved in 0c85346f85b9b2b0dfd1d3dac9d6ae4400e2e039

jeanmajid commented on behavior_pack/src/blockData/blockDataManager.ts:
```diff
@@ -0,0 +1,88 @@
+import { Block, ItemStack, world } from "@minecraft/server";
+
+const blockDataItemDynamicPropertyId = "blockData";
+
```
The whole block data system is actually pretty cool. Instead of having just one big giant dynamic propery, which is bad for perfomance and is just gonna cause issues down the road, try structuring it more like this "purpose:id" or in your case "blockData:blockLocation" so something like "blockData:12,32,-12".

BrodBlox09 replied:
❤️  Resolved in 38efe2a88bbca5.

jeanmajid commented on behavior_pack/src/components/blocks/fryingPan.ts:
```diff
@@ -3,36 +3,58 @@
 // See LICENSE.md file in the root folder, licenses/MIT.md, or https://opensource.org/license/mit
 
-import { BlockComponentPlayerInteractEvent } from "@minecraft/server";
+import { Block, BlockComponentPlayerInteractEvent, ContainerSlot, ItemStack, Player } from "@minecraft/server";
```
make sure you keep your imports clean, remove anything unused

BrodBlox09 replied:
Resolved in e2fa3da

jeanmajid reviewed (changes requested):
Some changed needed

BrodBlox09 committed:
Make more readable the setItem method calls that clear a slot

Instead of an empty space, specify undefined and add comments to further clarify.

BrodBlox09 committed:
Merge branch 'fryingPan' of https://github.com/BrodBlox09/Cook-Off into fryingPan

BrodBlox09 committed:
Remove unused imports

BrodBlox09 committed:
Break up blockData dynamic property

BlockData dynamic property broken into smaller dynamic properties for each block.
Also improve type safety by tagging BlockDataKey so not just any ol' string suffices.

BrodBlox09 committed:
Use existing item component system for admin tools

Admin tools are now used by using the admin tools item instead of sending a script event.

BrodBlox09 committed:
TSC

BrodBlox09 committed:
Enable strict type-checking

BrodBlox09 committed:
Resolve all type errors

jeanmajid reviewed (changes requested):
please fix this last thing

BrodBlox09 committed:
Merge branch 'main' into fryingPan

BrodBlox09 commented:
@jeanmajid that took a long time to show up :/ resolved
